### PR TITLE
FT2Font extension improvements

### PR DESCRIPTION
**Pull request by QuLogic**
## PR summary

This does two things:
1. Add more documentation to all of `ft2font`, and follow numpydoc style.
2. Convert FreeType enums/flags from "just a bunch of constants" to `Enum`-like objects. ~~Unfortunately, `pybind11` doesn't support Python's new-ish `Enum` class, but it should be a reasonable approximation.~~ This is based on the implementation in #9763 and mplcairo, but extended to all the flags as well.
   The `Enum` change still accepts plain integers, but raises a deprecation warning (as does accessing those constants from the module.) ~~I have not written an API change note in case we don't want to do this.~~ We may also want to mark the deprecation as pending.

## PR checklist

- [n/a] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [n/a] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [x] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [x] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

**Comment by anntzer** [comment]
If you want real Pythonic enums (that actually inherit from enum.Enum) you can reuse the P11X_DECLARE_ENUM macro I wrote in mplcairo.

**Comment by QuLogic** [comment]
I can't say I understand all of that macro, but I can certainly copy it over if that's the way we want to go.

**Comment by anntzer** [comment]
(I don't really mind either way, but you mentioned that real enums may be better, so I suggested it.  I can also try to explain better how it works...)

**Comment by QuLogic** [comment]
The failure in the docs actually shows the confusion with the "just a bunch of constants" approach:
https://github.com/matplotlib/matplotlib/blob/96751227d9e3f8f8440b425d0916e593c17667c3/galleries/examples/misc/ftface_props.py#L49-L63
Only Bold and Italic are style flags; the rest are face flags, and checking them as they are here would always return False.

**Comment by QuLogic** [comment]
> (I don't really mind either way, but you mentioned that real enums may be better, so I suggested it. I can also try to explain better how it works...)

On second reading, I have some understanding of it all. It would be nice though if I could add a docstring to each class (as done here), and avoid the C++ enum (but I think I need this one for the `variant` to produce the deprecation anyway.)

**Comment by QuLogic** [comment]
> I guess to add a docstring you need to do something like `p11x::enums[py_name].attr("__doc__") = ...;` as there's no way to set it in the function-style API.

Yep, looks like that worked nicely.

> The macro is explicitly designed to interface with a C(++) enum but you can probably get away without it if you also remove the entire typecaster machinery?

I started making a copy without the `is_enum_v` and hardcoding instead of `underlying_type_t`, but realized I still needed the enum for the deprecation, I think, so gave up on that. Would be nice to de-dupe, but at least the two copies are next to each other now.

File: doc/api/next_api_changes/deprecations/28842-ES.rst [hunk 1]
```diff
@@ -0,0 +1,53 @@
+ft2font module-level constants replaced by enums
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The `.ft2font`-level constants have been converted to `enum` classes, and all API using
+them now take/return the new types.
+
+The following constants are now part of `.ft2font.Kerning`:
+
+- ``KERNING_DEFAULT``
```
Comment (anntzer) [inline doc/api/next_api_changes/deprecations/28842-ES.rst:9]: perhaps note that the new name doesn't include KERNING_; ditto for LOAD_.

File: src/_enums.h [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    // converted to the Python class object after init.
+    auto enums = std::unordered_map<std::string, py::object>{};
+
+    auto bind_enums(py::module mod) -> void
+    {
+      for (auto& [py_name, spec]: enums) {
+        auto const& [py_base_cls, pairs] =
+          spec.cast<std::pair<std::string, py::object>>();
+        mod.attr(py::cast(py_name)) = spec =
+          py::module::import("pydoc").attr("locate")(py_base_cls)(
```
Comment (anntzer) [inline src/_enums.h:30]: You can probably use pkgutil.resolve_name here instead of the older and undocumented pydoc.locate (though both would work).

Comment (QuLogic) [reply src/_enums.h:30]: Or should we just hard-code this coming from `enum` (and only provide the class name to the macro)?

Comment (anntzer) [reply src/_enums.h:30]: That's probably fine too, anything works here...

File: doc/api/next_api_changes/deprecations/28842-ES.rst (hunk 1 above)
Comment (QuLogic) [reply doc/api/next_api_changes/deprecations/28842-ES.rst:10]: Done.

File: src/_enums.h (hunk 2 above)
Comment (QuLogic) [reply src/_enums.h]: OK, I just did the hard-coding then.

**Comment by QuLogic** [comment]
Ah, apparently using `#ifdef` in macro calls is undefined, and MSVC doesn't accept it. I'll just comment out the new ones until we update FreeType.

**Comment by QuLogic** [comment]
If I've understood the remaining failure, MSVC doesn't expand `__VA_ARGS__` in the `P11X_ENUM_TYPE` macro call to be multiple arguments, so everything falls apart after. I worked around that by just manually passing the type, unless you have any better ideas?

**Comment by anntzer** [comment]
> If I've understood the remaining failure, MSVC doesn't expand `__VA_ARGS__` in the `P11X_ENUM_TYPE` macro call to be multiple arguments, so everything falls apart after. I worked around that by just manually passing the type, unless you have any better ideas?

You likely need to pass `/Zc:preprocessor` (or the older equivalent `/experimental:preprocessor`, which is what I use in mplcairo) (https://learn.microsoft.com/en-us/cpp/preprocessor/preprocessor-experimental-overview?view=msvc-170) in extra_compile_args (or whatever that's called for meson).

**Comment by QuLogic** [comment]
OK, I checked and `/experimental:preprocessor` should be old enough for us to depend on.

**Comment by QuLogic** [comment]
So this is failing on AppVeyor because we're still on the 2017 image; I opened #28869 separately to see if we can bump to the 2019 one.

**Comment by QuLogic** [comment]
We're passing now that AppVeyor has been updated.

**Comment by tacaswell** [comment]
sorry, took two tries to get the rebase right.

**Review by greglucas — APPROVED** [review]

**Review by tacaswell — APPROVED** [review]

**Pull request merged**